### PR TITLE
update push_status when incremental EDS

### PR DESCRIPTION
**Pull request by hzxuzhonghu**

### CODE REVIEW

File: pilot/pkg/proxy/envoy/v2/discovery.go [hunk 1]
```diff
@@ -311,16 +311,18 @@ func (s *DiscoveryServer) periodicRefreshMetrics() {
...
 	if pc != nil {
 		pc.OnConfigChange()
 	}
+
+	if !full {
+		adsLog.Infof("XDS Incremental Push EDS:%d", len(edsUpdates))
+		go s.AdsPushAll(versionInfo(), s.globalPushContext(), false, edsUpdates)
+		return
+	}
+
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/discovery.go]: Why this change? and what is it doing? this whole code is super hacky :(

File: pilot/pkg/model/push_context.go [hunk 2]
```diff
@@ -168,6 +168,23 @@ func (ps *PushContext) Add(metric *PushMetric, key string, proxy *Proxy, msg str
 	metricMap[key] = ev
 }
 
+// Delete will remove an case from the metric.
+func (ps *PushContext) Delete(metric *PushMetric, key string) {
```
Comment (rshriram) [inline pilot/pkg/model/push_context.go]: can you rename all these functions to MetricAddEndpoint MetricDeleteEndpoint, etc. ? So that its easier to distinguish between metrics vs actual add/delete?

File: pilot/pkg/proxy/envoy/v2/eds.go [hunk 3]
```diff
@@ -258,6 +258,8 @@ func (s *DiscoveryServer) updateClusterInc(push *model.PushContext, clusterName
 	if cnt == 0 {
 		push.Add(model.ProxyStatusClusterNoInstances, clusterName, nil, "")
 		//adsLog.Infof("EDS: no instances %s (host=%s ports=%v labels=%v)", clusterName, hostname, p, labels)
+	} else {
+		push.Delete(model.ProxyStatusClusterNoInstances, clusterName)
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/eds.go]: See previous comment. this reads as if count==0, add somecluster, else delete some cluster.. which is not what the code is doing.

File: pilot/pkg/proxy/envoy/v2/discovery.go [hunk 4]
```diff
@@ -311,16 +311,18 @@ func (s *DiscoveryServer) periodicRefreshMetrics() {
...
 func (s *DiscoveryServer) Push(full bool, edsUpdates map[string]*EndpointShardsByService) {
-	if !full {
-		adsLog.Infof("XDS Incremental Push EDS:%d", len(edsUpdates))
-		go s.AdsPushAll(versionInfo(), s.globalPushContext(), false, edsUpdates)
-		return
-	}
 	// Reset the status during the push.
 	pc := s.globalPushContext()
 	if pc != nil {
 		pc.OnConfigChange()
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/discovery.go]: isn't this going to cause a lot of config recomputation etc. ?

File: pilot/pkg/model/push_context.go (hunk 2 above)
Comment (costinm) [reply pilot/pkg/model/push_context.go]: +1 ( but can be separate PR)

Comment (costinm) [reply pilot/pkg/model/push_context.go]: Also not clear we need to delete - just set to 0.

File: pilot/pkg/proxy/envoy/v2/discovery.go [hunk 5]
```diff
@@ -311,16 +311,18 @@ func (s *DiscoveryServer) periodicRefreshMetrics() {
...
 	}
+
+	if !full {
```
Comment (costinm) [inline pilot/pkg/proxy/envoy/v2/discovery.go]: I don't think EDS change should cause onConfigChange() - just needs to update metrics related to clusters.

Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/discovery.go]: This is `OnConfigChange` does now, set pushStatus and update metrics.
```
// OnConfigChange is called when a config change is detected.
func (ps *PushContext) OnConfigChange() {
	LastPushMutex.Lock()
	LastPushStatus = ps
	LastPushMutex.Unlock()
	ps.UpdateMetrics()
}
```

Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/discovery.go]: Only by this we can get the accurate status info from debug endpoint.

File: pilot/pkg/model/push_context.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go]: Actually this is not metrics, it is `  ProxyStatus map[string]map[string]ProxyPushStatus` . And 

`PushContext.UpdateMetrics` set metrics by counting `ProxyStatus ` fields.  So I can change the comments.

File: pilot/pkg/proxy/envoy/v2/discovery.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/discovery.go]: Just move push status reset in front of  incremental eds.
